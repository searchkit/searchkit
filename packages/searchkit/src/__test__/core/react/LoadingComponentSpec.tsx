import * as React from "react";

import {
  SearchkitManager,
  LoadingComponent
} from "../../../"

import {mount} from "enzyme";

describe("SearchkitProvider", ()=> {

  beforeEach(()=> {
    this.searchkit = SearchkitManager.mock()
    this.wrapper = mount(
      <LoadingComponent searchkit={this.searchkit}>
        <p>loading...</p>
      </LoadingComponent>
    )
  })

  it("display nothing when not loading", ()=> {
    expect(this.wrapper).toMatchSnapshot()
  })

  it("display children when loading", ()=> {
    this.searchkit.loading = true
    this.wrapper.instance().forceUpdate()
    expect(this.wrapper).toMatchSnapshot()
  })


})
