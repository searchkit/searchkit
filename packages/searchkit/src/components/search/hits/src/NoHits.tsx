import * as PropTypes from "prop-types";

import {
	SearchkitComponent,
	SearchkitComponentProps,
	NoFiltersHitCountAccessor,
	SuggestionsAccessor,
	RenderComponentType,
	renderComponent
} from "../../../../core"

import { NoHitsErrorDisplay, NoHitsErrorDisplayProps } from "./NoHitsErrorDisplay"
import { NoHitsDisplay, NoHitsDisplayProps } from "./NoHitsDisplay"

const defaults = require("lodash/defaults")

export interface NoHitsProps extends SearchkitComponentProps {
	suggestionsField?: string
	errorComponent?: RenderComponentType<NoHitsErrorDisplayProps>
	component?: RenderComponentType<NoHitsDisplayProps>
}

export class NoHits extends SearchkitComponent<NoHitsProps, any> {
	noFiltersAccessor: NoFiltersHitCountAccessor
	suggestionsAccessor: SuggestionsAccessor
	bemBlocks: {
		container: Function
	}

	static translations = {
		"NoHits.NoResultsFound": "No results found for {query}.",
		"NoHits.NoResultsFoundDidYouMean": "No results found for {query}. Did you mean {suggestion}?",
		"NoHits.DidYouMean": "Search for {suggestion} instead",
		"NoHits.SearchWithoutFilters": "Search for {query} without filters",
		"NoHits.Error": "We're sorry, an issue occurred when fetching your results. Please try again.",
		"NoHits.ResetSearch": "Reset Search"
	}
	translations = NoHits.translations

	static propTypes = defaults({
		suggestionsField: PropTypes.string,
		translations: SearchkitComponent.translationsPropType(
			NoHits.translations
		)
	}, SearchkitComponent.propTypes)

	static defaultProps = {
		errorComponent: NoHitsErrorDisplay,
		component: NoHitsDisplay
	}

	componentDidMount() {
		super.componentDidMount()
		this.noFiltersAccessor = this.searchkit.addAccessor(
			new NoFiltersHitCountAccessor()
		)
		if (this.props.suggestionsField) {
			this.suggestionsAccessor = this.searchkit.addAccessor(
				new SuggestionsAccessor(this.props.suggestionsField)
			)
		}
	}

	defineBEMBlocks() {
		let block = (this.props.mod || "sk-no-hits")
		return {
			container: block
		}
	}

	getSuggestion() {
		return this.suggestionsAccessor && this.suggestionsAccessor.getSuggestion()
	}

	setQueryString(query) {
		this.searchkit.getQueryAccessor().setQueryString(query, true)
		this.searchkit.performSearch(true)
	}

	resetFilters() {
		this.searchkit.getQueryAccessor().keepOnlyQueryState()
		this.searchkit.performSearch(true)
	}

	resetSearch() {
		this.searchkit.getQueryAccessor().resetState()
		this.searchkit.performSearch(true)
	}

	getFilterCount() {
		return this.noFiltersAccessor && this.noFiltersAccessor.getCount()
	}

	render() : React.ReactElement<any> {
		if ((this.hasHits() || this.isInitialLoading() || this.isLoading()) && !this.getError()) return null

		if (this.getError()) {
			const props: NoHitsErrorDisplayProps = {
				errorLabel: this.translate("NoHits.Error"),
				resetSearchFn: this.resetSearch.bind(this),
				translate: this.translate,
				bemBlocks: this.bemBlocks,
				tryAgainLabel: this.translate("NoHits.ResetSearch"),
				error: this.getError()
			}
			return renderComponent(this.props.errorComponent, props)
		}

		const suggestion: any = this.getSuggestion()
		const query = this.getQuery().getQueryString()
		let infoKey = suggestion ? "NoHits.NoResultsFoundDidYouMean" : "NoHits.NoResultsFound"

		const props: NoHitsDisplayProps = {
			noResultsLabel: this.translate(infoKey, { query: query, suggestion: suggestion }),
			translate: this.translate,
			bemBlocks: this.bemBlocks,
			suggestion: suggestion,
			query: query,
			filtersCount: this.getFilterCount(),
			resetFiltersFn: this.resetFilters.bind(this),
			setSuggestionFn: this.setQueryString.bind(this, suggestion)
		}

		return renderComponent(this.props.component, props)

	}
}
