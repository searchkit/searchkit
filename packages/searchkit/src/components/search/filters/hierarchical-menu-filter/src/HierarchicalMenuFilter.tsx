import * as React from "react";
import * as PropTypes from "prop-types";

import {
	SearchkitComponent,
	HierarchicalFacetAccessor,
	FastClick,
	SearchkitComponentProps,
	renderComponent,
	RenderComponentType,
	RenderComponentPropType
} from "../../../../../core"

import {
	Panel, ItemComponent, ItemProps
} from "../../../../ui"

const defaults = require("lodash/defaults")
const map = require("lodash/map")
const identity = require("lodash/identity")

export interface HierarchicalMenuFilterProps extends SearchkitComponentProps {
	id: string
	fields: Array<string>
	title: string
	size?: number
	orderKey?: string
	orderDirection?: string
	countFormatter?: (count: number) => string | number
	containerComponent?: RenderComponentType<any>
	itemComponent?: RenderComponentType<ItemProps>
}

export class HierarchicalMenuFilter extends SearchkitComponent<HierarchicalMenuFilterProps, any> {
	public accessor: HierarchicalFacetAccessor

	static defaultProps = {
		countFormatter: identity,
		size: 20,
		containerComponent:Panel,
		itemComponent:ItemComponent
	}
	static propTypes = defaults({
		id: PropTypes.string.isRequired,
		fields: PropTypes.arrayOf(PropTypes.string).isRequired,
		title: PropTypes.string.isRequired,
		orderKey: PropTypes.string,
		orderDirection: PropTypes.oneOf(["asc", "desc"]),
		countFormatter: PropTypes.func,
		containerComponent: RenderComponentPropType,
		itemComponent: RenderComponentPropType
	}, SearchkitComponent.propTypes)

	defineBEMBlocks() {
		var blockClass = this.props.mod || "sk-hierarchical-menu";
		return {
			container: `${blockClass}-list`,
			option: `${blockClass}-option`
		};
	}

	defineAccessor() {
		const { id, title, fields, size, orderKey, orderDirection } = this.props
		return new HierarchicalFacetAccessor(id, {
			id, title, fields, size, orderKey, orderDirection
		})
	}

	addFilter(option, level) {
		this.accessor.state = this.accessor.state.toggleLevel(
			level, option.key)

		this.searchkit.performSearch()
	}

	renderOption(level, option) {

		var block = this.bemBlocks.option
		const { countFormatter, itemComponent } = this.props
		const active = this.accessor.state.contains(level, option.key) 
		
		return (
			<div key={option.key}>
				{renderComponent(itemComponent, {
					active,
					itemKey:option.key,
					showCount:true,
					bemBlocks:this.bemBlocks,
					onClick:this.addFilter.bind(this, option, level),
					label:this.translate(option.key),
					count:countFormatter(option.doc_count)
				})}
				{(() => {
					if (this.accessor.resultsState.contains(level, option.key)) {
						return this.renderOptions(level + 1);
					}
				})()}
			</div>
		)
	}

	renderOptions(level) {
		let block = this.bemBlocks.container;
		return (
			<div className={block("hierarchical-options")}>
				{map(this.accessor.getBuckets(level), this.renderOption.bind(this, level))}
			</div>
		)
	}

	render() {
		if(!this.accessor) return null;
		const block = this.bemBlocks.container
		const { id, title, containerComponent } = this.props
		return renderComponent(
			containerComponent, {
				title,
				className: id ? `filter--${id}` : undefined,
				disabled: this.accessor.getBuckets(0).length == 0
			}, 
				<div className={block("root")}>
					{this.renderOptions(0)}
				</div>
			
		)
	}
}
