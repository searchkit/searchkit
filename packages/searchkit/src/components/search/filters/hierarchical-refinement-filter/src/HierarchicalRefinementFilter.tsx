import * as React from "react"
import * as PropTypes from "prop-types"

import {
	SearchkitComponent,
	NestedFacetAccessor,
	FastClick,
	SearchkitComponentProps,
	RenderComponentType,
	renderComponent,
	RenderComponentPropType
} from "../../../../../core"

import {
	Panel, ItemComponent, ItemProps
} from "../../../../ui"

const defaults = require("lodash/defaults")
const map = require("lodash/map")
const identity = require("lodash/identity")

export interface HierarchicalRefinementFilterProps extends SearchkitComponentProps {
	field: string
	id: string
	title: string
	size?: number
	orderKey?: string
	orderDirection?: string
	startLevel?: number
	countFormatter?: (count: number) => number | string
	containerComponent?: RenderComponentType<any>
	itemComponent?: RenderComponentType<ItemProps>
}

export class HierarchicalRefinementFilter extends SearchkitComponent<HierarchicalRefinementFilterProps, any> {
	public accessor: NestedFacetAccessor

	static defaultProps = {
		countFormatter: identity,
		containerComponent: Panel,
		itemComponent: ItemComponent
	}

	static propTypes = defaults({
		field: PropTypes.string.isRequired,
		id: PropTypes.string.isRequired,
		title: PropTypes.string.isRequired,
		orderKey: PropTypes.string,
		orderDirection: PropTypes.oneOf(["asc", "desc"]),
		startLevel: PropTypes.number,
		countFormatter: PropTypes.func,
		containerComponent: RenderComponentPropType,
		itemComponent: RenderComponentPropType
	}, SearchkitComponent.propTypes)

	defineBEMBlocks() {
		var blockClass = this.props.mod || "sk-hierarchical-refinement";
		return {
			container: `${blockClass}-list`,
			option: `${blockClass}-option`
		}
	}

	defineAccessor() {
		const {
			field, id, title, size, orderKey,
			orderDirection, startLevel 
		} = this.props;

		return new NestedFacetAccessor(id, {
			field, id, title, size, orderKey,
			orderDirection, startLevel
		})
	}

	addFilter(level, option) {
		this.accessor.state = this.accessor.state.toggleLevel(level, option.key)
		this.searchkit.performSearch()
	}

	renderOption(level, option) {

		var block = this.bemBlocks.option
		var active = this.accessor.resultsState.contains(level, option.key)
		const { countFormatter, itemComponent } = this.props		

		return (
			<div key={option.key}>
				{renderComponent(itemComponent, {
					active,
					bemBlocks:this.bemBlocks,
					label:this.translate(option.key),
					itemKey:option.key,
					count:countFormatter(option.doc_count),
					showCount:true,
					onClick: this.addFilter.bind(this, level, option)
				})}
				{active && this.renderOptions(level + 1)}					
			</div>
		)
	}

	renderOptions(level) {
		let block = this.bemBlocks.container;
		return (
			<div className={block("hierarchical-options")}>
				{
					map(this.accessor.getBuckets(level), this.renderOption.bind(this, level))
				}
			</div>
		)
	}

	render() {
		if(!this.accessor) return null;
		const block = this.bemBlocks.container
		const { id, title, containerComponent } = this.props
		return renderComponent(
			containerComponent, {
				title,
				className: id ? `filter--${id}` : undefined,
				disabled: this.accessor.getBuckets(0).length == 0
			},
			<div className={block("root")}>
				{this.renderOptions(0)}
			</div>

		)
	}


}
