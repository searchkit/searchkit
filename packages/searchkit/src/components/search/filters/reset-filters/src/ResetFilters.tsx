import * as React from "react";
import * as PropTypes from "prop-types";

import {
	SearchkitComponent,
	FastClick,
	SearchkitComponentProps,
	RenderComponentType,
	ResetSearchOptions,
	ResetSearchAccessor,
  renderComponent
} from "../../../../../core"
const defaults = require("lodash/defaults")


export interface ResetFiltersDisplayProps {
	bemBlock:any,
	hasFilters:boolean,
	resetFilters:Function,
	clearAllLabel:string,
	translate:Function
}

export class ResetFiltersDisplay extends React.PureComponent<ResetFiltersDisplayProps, any>{
	render(){
		const { bemBlock, hasFilters, resetFilters, clearAllLabel } = this.props
		return (
			<div>
				<FastClick handler={resetFilters}>
					<div className={bemBlock().state({disabled:!hasFilters})}>
						<div className={bemBlock("reset")}>{clearAllLabel}</div>
					</div>
				</FastClick>
			</div>
		)
	}
}

export interface ResetFiltersProps extends SearchkitComponentProps {
	component?: RenderComponentType<ResetFiltersDisplayProps>,
	options?:ResetSearchOptions
}

export class ResetFilters extends SearchkitComponent<ResetFiltersProps, any> {

	static translations:any = {
		"reset.clear_all":"Clear all filters"
	}
	translations = ResetFilters.translations
	accessor:ResetSearchAccessor

	static propTypes = defaults({
		translations:SearchkitComponent.translationsPropType(
			ResetFilters.translations
		),
		component:PropTypes.func,
		options:PropTypes.object
	}, SearchkitComponent.propTypes)

	static defaultProps = {
		component:ResetFiltersDisplay
	}

	constructor(props){
		super(props)
		this.resetFilters = this.resetFilters.bind(this)
	}

	defineBEMBlocks() {
		return {
			container: (this.props.mod || "sk-reset-filters")
		}
	}

	defineAccessor(){
		return new ResetSearchAccessor(this.props.options)
	}

	resetFilters() {
		this.accessor.performReset()
		this.searchkit.performSearch()
	}

  render() {
	if(!this.accessor) return null;
		const props:ResetFiltersDisplayProps = {
			bemBlock:this.bemBlocks.container,
			resetFilters:this.resetFilters,
			hasFilters:this.accessor.canReset(),
			translate:this.translate,
			clearAllLabel: this.translate("reset.clear_all")
		}
		return renderComponent(this.props.component, props)
  }
}
