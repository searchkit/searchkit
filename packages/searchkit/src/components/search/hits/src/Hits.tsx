import * as React from "react";
import * as PropTypes from "prop-types";

import {
	SearchkitComponent,
	PageSizeAccessor,
	HighlightAccessor,
	CustomHighlightAccessor,
	SearchkitComponentProps,
	SourceFilterType,
	SourceFilterAccessor,
	HitsAccessor,
	RenderComponentType,
	RenderComponentPropType,
	renderComponent,
	block
} from "../../../../core"

const map = require("lodash/map")
const defaults = require("lodash/defaults")



export interface HitItemProps {
	key:string,
	bemBlocks?:any,
	result:any
}

export class HitItem extends React.PureComponent<HitItemProps, any> {

	render(){
		return (
			<div data-qa="hit"
				className={this.props.bemBlocks.item().mix(this.props.bemBlocks.container("item"))}>
				{this.props.result._id}
			</div>
		)
	}
}

export interface HitsListProps{
	mod?:string,
	className?:string,
	itemComponent?:RenderComponentType<HitItemProps>,
	hits:Array<Object>
}

export class HitsList extends React.PureComponent<HitsListProps, any>{

	static defaultProps={
		mod:"sk-hits",
		itemComponent:HitItem
	}

	static propTypes = {
		mod:PropTypes.string,
		className:PropTypes.string,
		itemComponent:RenderComponentPropType,
		hits:PropTypes.any
	}

	render(){
		const {hits, mod, className, itemComponent} = this.props
		const bemBlocks = {
			container: block(mod).el,
			item: block(`${mod}-hit`).el
		}
		return (
			<div data-qa="hits" className={bemBlocks.container().mix(className)}>
				{map(hits, (result: any, index)=> {
					return renderComponent(itemComponent, {
						key:result._id, result, bemBlocks, index
					})
				})}
			</div>
		)
	}
}

export interface HitsProps extends SearchkitComponentProps{
	hitsPerPage?: number
	highlightFields?:Array<string>
	customHighlight?:any
	sourceFilter?:SourceFilterType
	itemComponent?: RenderComponentType<HitItemProps>
	listComponent?: RenderComponentType<HitsListProps>
	scrollTo?: boolean|string
}


export class Hits extends SearchkitComponent<HitsProps, any> {
	hitsAccessor:HitsAccessor

	static propTypes = defaults({
		hitsPerPage:PropTypes.number,
		highlightFields:PropTypes.arrayOf(
			PropTypes.string
		),
		sourceFilterType:PropTypes.oneOf([
			PropTypes.string,
			PropTypes.arrayOf(PropTypes.string),
			PropTypes.bool
		]),
		itemComponent:RenderComponentPropType,
		listComponent:RenderComponentPropType
	}, SearchkitComponent.propTypes)

	static defaultProps = {
		listComponent:HitsList,
		scrollTo: "body"
	}

	componentDidMount() {
		super.componentDidMount()
		if(this.props.hitsPerPage){
			this.searchkit.getAccessorByType(PageSizeAccessor)
				.defaultSize = this.props.hitsPerPage
		}
		if(this.props.highlightFields) {
			this.searchkit.addAccessor(
				new HighlightAccessor(this.props.highlightFields))
		}
		if (this.props.customHighlight) {
			this.searchkit.addAccessor(new CustomHighlightAccessor(this.props.customHighlight))
		}
		if(this.props.sourceFilter){
			this.searchkit.addAccessor(
				new SourceFilterAccessor(this.props.sourceFilter)
			)
		}
		this.hitsAccessor = new HitsAccessor({ scrollTo:this.props.scrollTo })
		this.searchkit.addAccessor(this.hitsAccessor)
	}


	render() {
		let hits:Array<Object> = this.getHits()
		let hasHits = hits.length > 0

		if (!this.isInitialLoading() && hasHits) {
			const {listComponent, mod, className, itemComponent} = this.props
			return renderComponent(listComponent, {
				hits, mod, className, itemComponent
			})
		}

		return null

	}
}
