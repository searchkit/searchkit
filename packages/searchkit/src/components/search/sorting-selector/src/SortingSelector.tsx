import * as PropTypes from "prop-types";
import * as React from 'react';

import {
	SearchkitComponent,
	SortingAccessor,
	SearchkitComponentProps,
	SortingOption,
	renderComponent,
	RenderComponentType,
	RenderComponentPropType
} from "../../../../core"

import {
	Select, ListProps
} from "../../../ui"

const defaults = require("lodash/defaults")

export interface SortingProps extends SearchkitComponentProps {
  options:Array<SortingOption>
  listComponent?: RenderComponentType<ListProps>
}

export class SortingSelector extends SearchkitComponent<SortingProps, any> {
  accessor:SortingAccessor

  static propTypes = defaults({
    listComponent: RenderComponentPropType,
    options:PropTypes.arrayOf(
      PropTypes.shape({
        label:PropTypes.string.isRequired,
        field:PropTypes.string,
        order:PropTypes.string,
        defaultOption:PropTypes.bool
      })
    )
  }, SearchkitComponent.propTypes)

  static defaultProps = {
    listComponent: Select
  }

  defineAccessor() {
    return new SortingAccessor("sort", {options:this.props.options})
  }

  toggleItem(key) {
    this.accessor.state = this.accessor.state.setValue(key);
    this.searchkit.performSearch();
  }

	setItems(keys){
		this.toggleItem(keys[0])
	}

  render(): React.ReactElement<any> {
    if (!this.accessor) return null;
    const { listComponent } = this.props
		const options = this.accessor.options.options
    const selected = [this.accessor.getSelectedOption().key]
    const disabled = !this.hasHits()

    return renderComponent(listComponent, {
			mod:this.props.mod,
			className:this.props.className,
      items: options,
      selectedItems: selected,
      setItems: this.setItems.bind(this),
      toggleItem: this.toggleItem.bind(this),
      disabled: disabled,
      urlBuilder: (item) => this.accessor.urlWithState(item.key),
			translate:this.translate
    })
  }
}
