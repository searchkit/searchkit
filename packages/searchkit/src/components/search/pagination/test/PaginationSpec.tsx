import * as React from "react";
import {mount} from "enzyme";
import {Pagination, PaginationSelect} from "../src/Pagination";
import {SearchkitManager, ImmutableQuery} from "../../../../core";
import {Select} from "../../../ui";
import {
  fastClick
} from "../../../__test__/TestHelpers"
import * as sinon from "sinon";

describe("Pagination tests", () => {

  beforeEach(() => {

    this.searchkit = SearchkitManager.mock()
    this.searchkit.addDefaultQuery((query)=> {
      return query.setSize(10)
    })
    this.createWrapper = (showNumbers=true, pageScope=3, props={}) => {
      this.wrapper = mount(
        <Pagination searchkit={this.searchkit}
                    showNumbers={showNumbers}
                    pageScope={pageScope}
                    {...props}
                    translations={{ "pagination.previous": "Previous Page" }} />
      );
      this.accessor = this.searchkit.accessors.statefulAccessors["p"]
    }

    this.searchkit.query = new ImmutableQuery().setSize(10)


    this.searchkit.setResults({
      hits:{
        total:80
      }
    })


  });

  describe("rendering", () => {

    beforeEach(() => {

      this.checkActionStates = (page) => {
        this.accessor.state = this.accessor.state.setValue(page)
        this.wrapper.update();
      }
    })

    it("renders text", () => {
      this.createWrapper()
      expect(this.wrapper).toMatchSnapshot()
    })

    it("renders with pages", () => {
      this.createWrapper(false)
      expect(this.wrapper).toMatchSnapshot()
    })

    it('renders first page options', () => {
      this.createWrapper()
      this.checkActionStates(
        null
      )
      this.createWrapper()
      expect(this.wrapper).toMatchSnapshot("first page options")
    })

    it('renders second page options', () => {
      this.createWrapper()
      this.checkActionStates(
        2)
      this.createWrapper()
      expect(this.wrapper).toMatchSnapshot("second page options")
    })

    it('renders eighth page options', () => {
      this.createWrapper()
      this.checkActionStates(
        8
      )
      this.createWrapper()
      expect(this.wrapper).toMatchSnapshot("eighth page options")
    })

    it("handles showNumbers prop", () => {
      this.createWrapper(false)
      this.checkActionStates(4)
      this.createWrapper(false)
      expect(this.wrapper).toMatchSnapshot("show numbers props")
    })

    it("handles pageScope prop", () => {
      this.createWrapper(true, 1)
      this.checkActionStates(
        4
      )
      this.createWrapper(true, 1)
      expect(this.wrapper).toMatchSnapshot("page scope prop")
    })

    it("renders no pagination on no results", () => {
      this.searchkit.setResults({hits:{total:0}})
      this.createWrapper()
      expect(this.wrapper.find(".sk-toggle").length).toBe(0)
    })

    it("both disabled on only one total page", () => {
      this.searchkit.setResults({ hits: { total: 10 } })
      this.createWrapper()
      this.checkActionStates(1)
      expect(this.wrapper).toMatchSnapshot("both disabled on only one total page")

    })

  });

  describe("interacting", () => {

    it("interact prev disabled", () => {
      this.createWrapper()
      this.accessor.state = this.accessor.state.setValue(1)

      fastClick(this.wrapper.find(".sk-toggle__item").first())
      expect(this.accessor.state.getValue()).toBe(1)
    });

    it("click previous, next", ()=> {
      this.createWrapper()
      this.accessor.state = this.accessor.state.setValue(3)
      this.wrapper.update()
      fastClick(this.wrapper.find(".sk-toggle__item").first())
      expect(this.accessor.state.getValue()).toBe(2)
      fastClick(this.wrapper.find(".sk-toggle__item").last())
      fastClick(this.wrapper.find(".sk-toggle__item").last())
      expect(this.accessor.state.getValue()).toBe(4)
    })

    it("ability to click last page", ()=> {
      this.createWrapper()
      this.accessor.state = this.accessor.state.setValue(7)
      this.wrapper.update()
      fastClick(this.wrapper.find(".sk-toggle__item").last())
      expect(this.accessor.state.getValue()).toBe(8)
      fastClick(this.wrapper.find(".sk-toggle__item").last())
      expect(this.accessor.state.getValue()).toBe(8)
    })

    it("dividers should not alter state", ()=> {
      this.createWrapper()
      this.accessor.state = this.accessor.state.setValue(2)
      this.wrapper.update()
      this.createWrapper()      
      fastClick(this.wrapper.find("[data-key='ellipsis-6']"))
      //this was NaN before bug fix
      expect(this.accessor.state.getValue()).toBe(2)
    })

  })

  it("PaginationSelect", () => {
    this.wrapper = mount(
      <PaginationSelect searchkit={this.searchkit} />
    );
    this.wrapper.update()
    expect(this.wrapper).toMatchSnapshot()
  })

});
