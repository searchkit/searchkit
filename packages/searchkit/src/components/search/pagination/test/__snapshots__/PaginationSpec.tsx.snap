// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination tests PaginationSelect 1`] = `
<div class="sk-pagination-select is-numbered">
  <div class="sk-select"><select><option value="1">1</option><option value="2">2</option><option value="3">3</option><option value="4">4</option></select></div>
</div>
`;

exports[`Pagination tests rendering renders text 1`] = `
<div class="sk-pagination-navigation is-numbered">
  <div data-qa="options" class="sk-toggle">
    <div class="sk-toggle-option sk-toggle__item is-disabled" data-qa="option" data-key="previous">
      <div data-qa="label" class="sk-toggle-option__text">Previous Page</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item is-active" data-qa="option" data-key="1">
      <div data-qa="label" class="sk-toggle-option__text">1</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item" data-qa="option" data-key="2">
      <div data-qa="label" class="sk-toggle-option__text">2</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item" data-qa="option" data-key="3">
      <div data-qa="label" class="sk-toggle-option__text">3</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item" data-qa="option" data-key="4">
      <div data-qa="label" class="sk-toggle-option__text">4</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item is-disabled" data-qa="option" data-key="ellipsis-5">
      <div data-qa="label" class="sk-toggle-option__text">...</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item" data-qa="option" data-key="next">
      <div data-qa="label" class="sk-toggle-option__text">Next</div>
    </div>
  </div>
</div>
`;

exports[`Pagination tests rendering renders with pages 1`] = `
<div class="sk-pagination-navigation">
  <div data-qa="options" class="sk-toggle">
    <div class="sk-toggle-option sk-toggle__item is-disabled" data-qa="option" data-key="previous">
      <div data-qa="label" class="sk-toggle-option__text">Previous Page</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item" data-qa="option" data-key="next">
      <div data-qa="label" class="sk-toggle-option__text">Next</div>
    </div>
  </div>
</div>
`;

exports[`both disabled on only one total page 1`] = `
<div class="sk-pagination-navigation is-numbered">
  <div data-qa="options" class="sk-toggle is-disabled">
    <div class="sk-toggle-option sk-toggle__item is-disabled" data-qa="option" data-key="previous">
      <div data-qa="label" class="sk-toggle-option__text">Previous Page</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item is-active" data-qa="option" data-key="1">
      <div data-qa="label" class="sk-toggle-option__text">1</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item is-disabled" data-qa="option" data-key="next">
      <div data-qa="label" class="sk-toggle-option__text">Next</div>
    </div>
  </div>
</div>
`;

exports[`eighth page options 1`] = `
<div class="sk-pagination-navigation is-numbered">
  <div data-qa="options" class="sk-toggle">
    <div class="sk-toggle-option sk-toggle__item" data-qa="option" data-key="previous">
      <div data-qa="label" class="sk-toggle-option__text">Previous Page</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item" data-qa="option" data-key="1">
      <div data-qa="label" class="sk-toggle-option__text">1</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item is-disabled" data-qa="option" data-key="ellipsis-2">
      <div data-qa="label" class="sk-toggle-option__text">...</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item" data-qa="option" data-key="5">
      <div data-qa="label" class="sk-toggle-option__text">5</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item" data-qa="option" data-key="6">
      <div data-qa="label" class="sk-toggle-option__text">6</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item" data-qa="option" data-key="7">
      <div data-qa="label" class="sk-toggle-option__text">7</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item is-active" data-qa="option" data-key="8">
      <div data-qa="label" class="sk-toggle-option__text">8</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item is-disabled" data-qa="option" data-key="next">
      <div data-qa="label" class="sk-toggle-option__text">Next</div>
    </div>
  </div>
</div>
`;

exports[`first page options 1`] = `
<div class="sk-pagination-navigation is-numbered">
  <div data-qa="options" class="sk-toggle">
    <div class="sk-toggle-option sk-toggle__item is-disabled" data-qa="option" data-key="previous">
      <div data-qa="label" class="sk-toggle-option__text">Previous Page</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item is-active" data-qa="option" data-key="1">
      <div data-qa="label" class="sk-toggle-option__text">1</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item" data-qa="option" data-key="2">
      <div data-qa="label" class="sk-toggle-option__text">2</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item" data-qa="option" data-key="3">
      <div data-qa="label" class="sk-toggle-option__text">3</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item" data-qa="option" data-key="4">
      <div data-qa="label" class="sk-toggle-option__text">4</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item is-disabled" data-qa="option" data-key="ellipsis-5">
      <div data-qa="label" class="sk-toggle-option__text">...</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item" data-qa="option" data-key="next">
      <div data-qa="label" class="sk-toggle-option__text">Next</div>
    </div>
  </div>
</div>
`;

exports[`page scope prop 1`] = `
<div class="sk-pagination-navigation is-numbered">
  <div data-qa="options" class="sk-toggle">
    <div class="sk-toggle-option sk-toggle__item" data-qa="option" data-key="previous">
      <div data-qa="label" class="sk-toggle-option__text">Previous Page</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item" data-qa="option" data-key="1">
      <div data-qa="label" class="sk-toggle-option__text">1</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item is-disabled" data-qa="option" data-key="ellipsis-2">
      <div data-qa="label" class="sk-toggle-option__text">...</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item" data-qa="option" data-key="3">
      <div data-qa="label" class="sk-toggle-option__text">3</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item is-active" data-qa="option" data-key="4">
      <div data-qa="label" class="sk-toggle-option__text">4</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item" data-qa="option" data-key="5">
      <div data-qa="label" class="sk-toggle-option__text">5</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item is-disabled" data-qa="option" data-key="ellipsis-6">
      <div data-qa="label" class="sk-toggle-option__text">...</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item" data-qa="option" data-key="next">
      <div data-qa="label" class="sk-toggle-option__text">Next</div>
    </div>
  </div>
</div>
`;

exports[`second page options 1`] = `
<div class="sk-pagination-navigation is-numbered">
  <div data-qa="options" class="sk-toggle">
    <div class="sk-toggle-option sk-toggle__item" data-qa="option" data-key="previous">
      <div data-qa="label" class="sk-toggle-option__text">Previous Page</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item" data-qa="option" data-key="1">
      <div data-qa="label" class="sk-toggle-option__text">1</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item is-active" data-qa="option" data-key="2">
      <div data-qa="label" class="sk-toggle-option__text">2</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item" data-qa="option" data-key="3">
      <div data-qa="label" class="sk-toggle-option__text">3</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item" data-qa="option" data-key="4">
      <div data-qa="label" class="sk-toggle-option__text">4</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item" data-qa="option" data-key="5">
      <div data-qa="label" class="sk-toggle-option__text">5</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item is-disabled" data-qa="option" data-key="ellipsis-6">
      <div data-qa="label" class="sk-toggle-option__text">...</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item" data-qa="option" data-key="next">
      <div data-qa="label" class="sk-toggle-option__text">Next</div>
    </div>
  </div>
</div>
`;

exports[`show numbers props 1`] = `
<div class="sk-pagination-navigation">
  <div data-qa="options" class="sk-toggle">
    <div class="sk-toggle-option sk-toggle__item" data-qa="option" data-key="previous">
      <div data-qa="label" class="sk-toggle-option__text">Previous Page</div>
    </div>
    <div class="sk-toggle-option sk-toggle__item" data-qa="option" data-key="next">
      <div data-qa="label" class="sk-toggle-option__text">Next</div>
    </div>
  </div>
</div>
`;
